### PR TITLE
Separated out upgrade from open.

### PR DESCRIPTION
**Pull request by ramanujam-raman**

### CODE REVIEW

File: core/backend/src/IModelDb.ts [hunk 1]
```diff
@@ -890,6 +890,59 @@ export abstract class IModelDb extends IModel {
     return nativeDb;
   }
 
+  /**
+   * Determines if the schemas in the Db must or can be upgraded by comparing them with those included in the
+   * current version of the software.
+   * @param pathName Path of the DgnDb file
```
Comment (kabentley) [inline core/backend/src/IModelDb.ts:896]: say "file name of the briefcase"

File: core/backend/src/BriefcaseManager.ts [hunk 2]
```diff
@@ -176,7 +161,6 @@ export class BriefcaseManager {
   public static get imodelClient(): IModelClient { return IModelHost.iModelClient; }
```
Comment (kabentley) [inline core/backend/src/BriefcaseManager.ts:161]: why does this method exist?

File: core/backend/src/IModelDb.ts [hunk 3]
```diff
@@ -2356,31 +2409,96 @@ export class BriefcaseDb extends IModelDb {
...
+    const nativeDb = this.openDgnDb({ path: briefcaseProps.fileName }, openMode, upgradeOptions);
+    assert(!nativeDb.hasUnsavedChanges(), "Expected schema upgrade to have saved any changes made");
+    const changeSetId = nativeDb.getParentChangeSetId();
+    if (!nativeDb.hasPendingTxns())
+      return changeSetId;
+    const reversedChangeSetId = nativeDb.getReversedChangeSetId();
+    assert(reversedChangeSetId === undefined, "Expected schema upgrade to have failed if there were reversed changes in the briefcase");
+
+    const token: IModelRpcProps = {
+      key: "",
```
Comment (kabentley) [inline core/backend/src/IModelDb.ts:2447]: use `IModelRpcOpenProps` and get rid of key:

File: core/backend/src/test/integration/IModelWrite.test.ts [hunk 4]
```diff
@@ -963,131 +966,87 @@ describe("IModelWriteTest (#integration)", () => {
...
+    const iModelId = await HubUtility.pushIModel(managerRequestContext, projectId, pathname, hubName, true);
 
-    /* User "manager" upgrades the briefcase */
+    // Download two copies of the briefcase - manager and super
     const args: RequestNewBriefcaseProps = {
       contextId: projectId,
       iModelId,
     };
-    await BriefcaseManager.downloadBriefcase(managerRequestContext, args);
+    const managerBriefcaseProps: LocalBriefcaseProps = await BriefcaseManager.downloadBriefcase(managerRequestContext, args);
```
Comment (kabentley) [inline core/backend/src/test/integration/IModelWrite.test.ts:989]: I dislike these unnecessary type annotations. It looks like you're trying to override the return type of the method.

File: core/backend/src/test/integration/IModelWrite.test.ts [hunk 5]
```diff
@@ -963,131 +966,87 @@ describe("IModelWriteTest (#integration)", () => {
...
+    const superBriefcaseProps: LocalBriefcaseProps = await BriefcaseManager.downloadBriefcase(superRequestContext, args);
+    superRequestContext.enter();
+
+    /* User "manager" upgrades the briefcase */
 
     // Validate the original state of the BisCore schema in the briefcase
-    let iModel: BriefcaseDb;
-    let result: DbResult = DbResult.BE_SQLITE_OK;
-    iModel = await BriefcaseDb.open(managerRequestContext, { fileName: args.fileName! });
+    let iModel: BriefcaseDb = await BriefcaseDb.open(managerRequestContext, { fileName: managerBriefcaseProps.fileName });
```
Comment (kabentley) [inline core/backend/src/test/integration/IModelWrite.test.ts:997]: ditto

File: core/common/src/BriefcaseTypes.ts [hunk 6]
```diff
@@ -166,3 +164,26 @@ export interface UpgradeOptions {
...
+ * Note: The state may vary depending on whether the Db is to be opened ReadOnly or ReadWrite. Generally the
+ * requirements for opening a Db ReadOnly are less stringent than opening it ReadWrite.
+ * @beta
+ */
+export enum SchemaState {
+  /** The schemas in the Db are up-to-date, and do not need to be upgraded before opening it with the current version of the software */
+  UpToDate,
+
+  /** It's required that the schemas in the Db be upgraded before it can be opened with the current version of the software */
+  UpgradeRequired,
```
Comment (kabentley) [inline core/common/src/BriefcaseTypes.ts:179]: even readonly? Please explain what you can do readonly vs. readwrite.

File: core/backend/src/IModelDb.ts (hunk 1 above)
Comment (ramanujam-raman) [reply core/backend/src/IModelDb.ts]: Fixed now - see new comment and argument name.

File: core/backend/src/BriefcaseManager.ts (hunk 2 above)
Comment (ramanujam-raman) [reply core/backend/src/BriefcaseManager.ts]: Not needed - removed now.

File: core/backend/src/test/integration/IModelWrite.test.ts (hunk 4 above)
Comment (ramanujam-raman) [reply core/backend/src/test/integration/IModelWrite.test.ts]: Removed this across all occurrences in the backend.

File: core/common/src/BriefcaseTypes.ts (hunk 6 above)
Comment (ramanujam-raman) [reply core/common/src/BriefcaseTypes.ts:181]: Added a few more comments now - please take a look.

File: core/backend/src/test/integration/IModelWrite.test.ts (hunk 5 above)
Comment (ramanujam-raman) [reply core/backend/src/test/integration/IModelWrite.test.ts]: Done.

File: core/backend/src/IModelDb.ts (hunk 3 above)
Comment (kabentley) [reply core/backend/src/IModelDb.ts]: i made several changes to ensure you don't return without closing the iModel. Please review.

**Review by kabentley — APPROVED** [review]

**Review by scsewall — APPROVED** [review]

**Review by calebmshafer — APPROVED** [review]

**Review by swwilson-bsi — APPROVED** [review]

**Pull request merged**